flow/multiagent_envs/traffic_light_grid.py

```diff
@@ -0,0 +1,260 @@
+
+These environments are used to train traffic lights to regulate traffic flow
+through an n x m grid.
+"""
+
+import numpy as np
+from gym.spaces.box import Box
+from gym.spaces.discrete import Discrete
+
+from flow.core import rewards
+from flow.envs.green_wave_env import PO_TrafficLightGridEnv
+from flow.multiagent_envs.multiagent_env import MultiEnv
+
+ADDITIONAL_ENV_PARAMS = {
+    # num of nearby lights the agent can observe {0, ..., num_traffic_lights-1}
+    "num_local_lights": 4,  # FIXME: not implemented yet
+    # num of nearby edges the agent can observe {0, ..., num_edges}
+    "num_local_edges": 4,  # FIXME: not implemented yet
+}
+
+
+class MultiTrafficLightGridPOEnv(PO_TrafficLightGridEnv, MultiEnv):
+    """Multiagent shared model version of PO_TrafficLightGridEnv.
+
+    Required from env_params: See parent class
+
+    States
+        See parent class
+    Actions
+        See parent class
+    Rewards
+        See parent class
+    Termination
+        See parent class
+
+    """
+
+    def __init__(self, env_params, sim_params, scenario, simulator='traci'):
+        super().__init__(env_params, sim_params, scenario, simulator)
+
+        for p in ADDITIONAL_ENV_PARAMS.keys():
+            if p not in env_params.additional_params:
+                raise KeyError(
+                    'Environment parameter "{}" not supplied'.format(p))
+
+        # number of nearest lights to observe, defaults to 4
+        self.num_local_lights = env_params.additional_params.get(
+            "num_local_lights", 4)
+
+        # number of nearest edges to observe, defaults to 4
+        self.num_local_edges = env_params.additional_params.get(
+            "num_local_edges", 4)
+
+    @property
+    def observation_space(self):
+        """State space that is partially observed.
+
+        Velocities, distance to intersections, edge number (for nearby
+        vehicles) from each direction, local edge information, and traffic
+        light state.
+        """
+        tl_box = Box(
+            low=0.,
+            high=1,
+            shape=(3 * 4 * self.num_observed +
+                   2 * self.num_local_edges +
+                   3 * (1 + self.num_local_lights),
+                   ),
+            dtype=np.float32)
+        return tl_box
+
+    @property
+    def action_space(self):
+        """See class definition."""
+        if self.discrete:
+            return Discrete(2)
+        else:
+            return Box(
+                low=-1,
+                high=1,
+                shape=(1,),
+                dtype=np.float32)
+
+    def get_state(self):
+        """Observations for each traffic light agent.
+
+        :return: dictionary which contains agent-wise observations as follows:
+        - For the self.num_observed number of vehicles closest and incoming
+        towards traffic light agent, gives the vehicle velocity, distance to
+        intersection, edge_number, density traffic light state.
```

Comment:
Nit: density, traffic

Reply:
Is this a duplicate? We give both the density for each vehicle and the density on the edge?

Reply:
jk, we don't give the density there.